ext/simdjson/src/bindings.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+  +----------------------------------------------------------------------+
+  | simdjson_php                                                         |
+  +----------------------------------------------------------------------+
+  | This source file is subject to version 2.0 of the Apache license,    |
+  | that is bundled with this package in the file LICENSE, and is        |
+  | available through the world-wide-web at the following url:           |
+  | http://www.apache.org/licenses/LICENSE-2.0.html                      |
+  +----------------------------------------------------------------------+
+  | Author: Jinxi Wang  <1054636713@qq.com>                              |
+  | Author: Máté Kocsis <kocsismate@woohoolabs.com>                      |
+  +----------------------------------------------------------------------+
+*/
+
+#include "php.h"
+#include "php_simdjson.h"
+
+#include "simdjson.h"
+#include "bindings.h"
+
+extern "C" {
+	#include <ext/json/php_json.h>
+	#include <Zend/zend_exceptions.h>
+}
+
+// see https://github.com/simdjson/simdjson/blob/master/doc/performance.md#reusing-the-parser-for-maximum-efficiency
+simdjson::dom::parser parser;
+
+WARN_UNUSED
+simdjson::error_code
+build_parsed_json_cust(
+	simdjson::dom::element &doc, const char *buf, size_t len, bool realloc_if_needed, u_short depth = simdjson::DEFAULT_MAX_DEPTH
+) {
+    auto error = parser.allocate(len, depth);
+
+    if (error) {
+        return error;
+    }
+
+    error = parser.parse(buf, len, realloc_if_needed).get(doc);
+    if (error) {
+        return error;
+    }
+
+    return simdjson::SUCCESS;
+}
+
+static zval create_array(simdjson::dom::element element)
+{
+    zval v;
+
+    switch (element.type()) {
+        //ASCII sort
+        case simdjson::dom::element_type::STRING :
+            ZVAL_STRING(&v, std::string_view(element).data());
+            break;
+        case simdjson::dom::element_type::INT64 : ZVAL_LONG(&v, int64_t(element));
+            break;
+        case simdjson::dom::element_type::UINT64 : ZVAL_LONG(&v, uint64_t(element));
+            break;
+        case simdjson::dom::element_type::DOUBLE : ZVAL_DOUBLE(&v, double(element));
+            break;
+        case simdjson::dom::element_type::BOOL :
+            ZVAL_BOOL(&v, bool(element));
+            break;
+        case simdjson::dom::element_type::NULL_VALUE :
+            ZVAL_NULL(&v);
+            break;
+        case simdjson::dom::element_type::ARRAY :
+            zval arr;
+            array_init(&arr);
```

Comment:
You might be able to speed the `::ARRAY` case up a bit more by imitating what array_values does.

Something along these lines - there's a need to check if the size is 0xFFFFFF in addition to that, though

```
  /**
   * Get the size of the array (number of immediate children).
   * It is a saturated value with a maximum of 0xFFFFFF: if the value
   * is 0xFFFFFF then the size is 0xFFFFFF or greater.
   */
  inline size_t size() const noexcept;
```

```c
	/* Initialize return array */
	array_init_size(arr, arrlen);
	zend_hash_real_init_packed(Z_ARRVAL_P(arr));

	/* Go through input array and add values to the return array */
	ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(arr)) {
            for (simdjson::dom::element child : simdjson::dom::array(element)) {
                zval value = create_array(child);
                ZEND_HASH_FILL_ADD(&value);
            }
	} ZEND_HASH_FILL_END();
```